File: Data Structures & Algorithms/reverse-a-linked-list/submission-1.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode reverseList(ListNode node) {
        if (node == null || node.next == null) return node;
        ListNode newNode = reverseList(node.next);

        node.next.next = node;
        node.next = null;
        return newNode;

    }
}

File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        # dummy, 3, 2, 1. we need to get the n + 1 node from end.
        dummy = ListNode(0, head)
        hi = dummy
        lo = dummy

        for _ in range(n + 1):
            hi = hi.next
        
        while hi:
            lo = lo.next
            hi = hi.next
        
        lo.next = lo.next.next
        return dummy.next